src/gpgpu-sim/gpu-sim.cc

```diff
@@ -885,6 +901,11 @@ unsigned gpgpu_sim::finished_kernel() {
 
 void gpgpu_sim::set_kernel_done(kernel_info_t *kernel) {
   unsigned uid = kernel->get_uid();
+  last_uid = uid;
+  unsigned long long streamID = kernel->get_streamid();
+  last_streamID = streamID;
+  gpu_kernel_time.at(streamID).at(uid).end_cycle = GLOBAL_TIMER;
+  printf("gpgpu_sim::set_kernel_done start_cycle=%llu, end_cycle=%llu\n", gpu_kernel_time.at(streamID).at(uid).start_cycle, gpu_kernel_time.at(streamID).at(uid).end_cycle);
```

Comment:
Why not just use the end_cycle in kernel_info_t? See ` kernel->end_cycle = gpu_sim_cycle + gpu_tot_sim_cycle;` below line 913

Reply:
Yea I remember this. The kernel info is deleted at this time. Then I would suggeat don't delete the kernel_info_t after a kernel finish. What I did was create a vector to record finished kernels and don't delete the pointer. You can do similar or something new if you want.

Reply:
Justin, there was a reason end_cycle wouldn't work right?  I think it was because the end_cycle count was dependent on when start_cycle started, and this wasn't stream safe or something like that?

Reply:
I thought Matt and I had a conversation on this before. I thought when the kernels overlap, we found the gpu_tot_sim_cycle is no longer accurate for this purpose so we added GLOBAL_TIMER?

Reply:
Yes yes. When a kernel is issued, the model will update the kernel_start_cycle and push the kernel_info_t into a vector to record that the kernel is running. When a kernel is finished, the vector is cleared. This is perfectly fine when there is only one kernel running. But in concurrent, assume you launched 2 kernel, then you add both kernel in to the vector. And then one kernel finish, and this is reset the vector. Then the kernel 2 will be re-launched. The kernel_start_cycle of the kernel 2 will be updated at this point. Which results in the behavior you described. 
My solution was don't clear the vector that records launched kernels. The launched kernel vector is only used for this purpose. So it should be fine. I can fix it after this is merged. I was just wondering if there was some other reason. As this is the reason, I know how to fix this. I will take care of that.